src/DCSAM.cpp

```diff
@@ -115,13 +119,17 @@ void DCSAM::updateDiscreteInfo(const gtsam::Values &continuousVals,
   if (continuousVals.empty()) return;
   // TODO(any): inefficient, consider storing indices of DCFactors
   // Update the DC factors with new continuous information.
-  for (size_t j = 0; j < dfg_.size(); j++) {
+  std::cout << "Discrete factors size: " << dcDiscreteFactors_.size() << std::endl;
+  for (auto factor : dcDiscreteFactors_) {
     boost::shared_ptr<DCDiscreteFactor> dcDiscrete =
-        boost::dynamic_pointer_cast<DCDiscreteFactor>(dfg_[j]);
+        boost::static_pointer_cast<DCDiscreteFactor>(factor);
     if (dcDiscrete) {
       dcDiscrete->updateContinuous(continuousVals);
       dcDiscrete->updateDiscrete(discreteVals);
     }
+    else {
+      std::cout << "something is wrong with DCDiscreteFactors" << std::endl;
+    }
```

Comment:
`dcDiscreteFactors_` is a vector of pointers to `DCDiscreteFactor`s, so I don't think the explicit check if `dcDiscrete` is null is needed here (as it was for the previous `dynamic_pointer_cast`. Feel free to check me on this, but I think the static pointer cast will only result in a `nullptr` if `factor` itself is a `nullptr`.